### PR TITLE
Feat/add validations to external secret data from remote ref

### LINKED ISSUE #2916: Add validations to ExternalSecretDataFromRemoteRef
**Is your feature request related to a problem? Please describe.**
Currently, we don't have any validations on [the ExternalSecretDataFromRemoteRef](https://github.com/external-secrets/external-secrets/blob/3599384660c5d4f51d1afd5f99dc9d9aa3a9a1ee/apis/externalsecrets/v1beta1/externalsecret_types.go#L258C6-L280), but while reading the source code I noticed [there are several pre-conditions we need to meet](https://github.com/external-secrets/external-secrets/blob/9ff86eab5114edca5449633927d7959039db53a1/pkg/controllers/externalsecret/externalsecret_controller_secret.go#L55-L61):

1. We should set either Find, Exact, or SourceRef & SourceRef.GeneratorRef. 
2. We should set SourceRef.GeneratorRef if we want to use SourceRef.

**Describe the solution you'd like**
Those conditions are not validated properly right now, so it would be better if we add validations [here](https://github.com/external-secrets/external-secrets/blob/8b0fa87f301abd5ac2d15a45493aa4609e433772/apis/externalsecrets/v1beta1/externalsecret_validator.go#L39).

### PR DESCRIPTION
**Pull request by LucasPimentel123**
## Problem Statement

Currently, there are no validations on the ExternalSecretDataFromRemoteRef, but it should
- set either Find, Exact, or SourceRef & SourceRef.GeneratorRef.
- set SourceRef.GeneratorRef if we want to use SourceRef.

## Related Issue

Relates to #2916 

## Proposed Changes

I added the validations on the validateExternalSecret function

## Checklist

- [x] I have read the [contribution guidelines](https://external-secrets.io/latest/contributing/process/#submitting-a-pull-request)
- [x] All commits are signed with `git commit --signoff`
- [x] My changes have reasonable test coverage
- [x] All tests pass with `make test`
- [x] I ensured my PR is ready for review with `make reviewable`

### CODE REVIEW

File: apis/externalsecrets/v1beta1/externalsecret_validator.go [hunk 1]
```diff
@@ -61,6 +61,14 @@ func validateExternalSecret(obj runtime.Object) (admission.Warnings, error) {
 		if findOrExtract && ref.SourceRef != nil && ref.SourceRef.GeneratorRef != nil {
 			errs = errors.Join(errs, fmt.Errorf("generator can not be used with find or extract"))
 		}
+
+		if ref.SourceRef != nil &&  ref.SourceRef.GeneratorRef == nil {
+			errs = errors.Join(errs, fmt.Errorf("generator must be set if SourceRef is used"))
+		}
+
+		if !findOrExtract && (ref.SourceRef != nil ||  ref.SourceRef.GeneratorRef != nil) {
```
Comment (shuheiktgw) [inline apis/externalsecrets/v1beta1/externalsecret_validator.go:69]: This condition seems to be wrong. It should be  

```
!findOrExtract && (ref.SourceRef == nil || ref.SourceRef.GeneratorRef == nil)
```

Also, I found that `!findOrExtract` is hard to understand. Why don't we just use the following condition?

```
ref.Find == nil && ref.Extract == nil && (ref.SourceRef == nil || ref.SourceRef.GeneratorRef == nil)
```

Comment (shuheiktgw) [reply apis/externalsecrets/v1beta1/externalsecret_validator.go:69]: Would you add test cases to cover the four cases?

File: apis/externalsecrets/v1beta1/externalsecret_validator.go [hunk 2]
```diff
@@ -61,6 +61,14 @@ func validateExternalSecret(obj runtime.Object) (admission.Warnings, error) {
...
+			errs = errors.Join(errs, fmt.Errorf("either find, extract or source and generator must be set"))
```
Comment (shuheiktgw) [inline apis/externalsecrets/v1beta1/externalsecret_validator.go:70]:
```suggestion
			errs = errors.Join(errs, fmt.Errorf("either extract, find, or sourceRef with generatorRef must be set to dataFrom"))
```

File: apis/externalsecrets/v1beta1/externalsecret_validator.go [hunk 3]
```diff
@@ -61,6 +61,14 @@ func validateExternalSecret(obj runtime.Object) (admission.Warnings, error) {
 		if findOrExtract && ref.SourceRef != nil && ref.SourceRef.GeneratorRef != nil {
 			errs = errors.Join(errs, fmt.Errorf("generator can not be used with find or extract"))
 		}
+
+		if ref.SourceRef != nil &&  ref.SourceRef.GeneratorRef == nil {
+			errs = errors.Join(errs, fmt.Errorf("generator must be set if SourceRef is used"))
```
Comment (shuheiktgw) [inline apis/externalsecrets/v1beta1/externalsecret_validator.go:66]:
```suggestion
			errs = errors.Join(errs, fmt.Errorf("generatorRef cannot be empty when using sourceRef in dataFrom"))
```

**Review by shuheiktgw — CHANGES_REQUESTED** [review]
I've left some comments, so please take a look at them!

**Comment by shuheiktgw** [comment]
Hi @LucasPimentel123 👋 Are you still working on this PR?

**Comment by gusfcarvalho** [comment]
Hi @LucasPimentel123 ! Are you still working on this PR?

**Comment by shuheiktgw** [comment]
I will take it over from here

**Comment by shuheiktgw** [comment]
Closing in favor of https://github.com/external-secrets/external-secrets/pull/3390

**Pull request closed without merging**